### ISSUE TITLE
Sequence constraint to enforce only unique items

### ISSUE THREAD

**Issue by uriyyo** [samuelcolvin/pydantic#2011]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this feature/change is needed
* [x] After submitting this, I commit to one of:
    * Look through open issues and helped at least one other person
    * Hit the "watch" button on this repo to receive notifications and I commit to help at least 2 people that ask questions in the future
    * Implement a Pull Request for a confirmed bug

# Feature Request

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.6.1
            pydantic compiled: False
                 install path: /Users/yuriikarabas/my-projects/pydantic/pydantic
               python version: 3.6.8 (v3.6.8:3c6b436a57, Dec 24 2018, 02:04:31)  [GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)]
                     platform: Darwin-19.6.0-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions', 'email-validator', 'devtools']

```

The idea of this feature is to add the ability to validate that sequence types (like `list`, `tuple`, or `set`) contains only unique items.

I think the best API for this feature will be something like this:
```python
from typing import List
from pydantic import BaseModel, EmailStr, Field

class User(BaseModel):
    email: EmailStr


class Form(BaseModel):
    users: List[User] = Field(unique=True)
```

In case when data contains duplicates, they will be available using `errors` method.
For instance:
```python
try:
    f = Form(users=[{'email': 'some@email.com'}] * 2)
except ValidationError as e:
    print(e.errors())
```
Will print:
```python
[
    {
        'loc': ('users',),
        'msg': "value contains duplicate items [User(email='some@email.com')]",
        'type': 'type_error.sequencenotunique',
        'ctx': {'duplicates': [User(email='some@email.com')]}
    },
]
```

Also, as an option, we can add a new generic type `Unique`. 
Usage will look like:
```python
from pydantic import BaseModel, EmailStr, Unique

class User(BaseModel):
    email: EmailStr


class Form(BaseModel):
    users: Unique[User]
```

I have already implemented this feature in my fork. [link to commit](https://github.com/uriyyo/pydantic/commit/bb67a2e9f15da0d826ba48da117230284182c1b3)

**Comment by mdavis-xyz** [comment]
As a workaround, maybe you could set the attribute type to a set? Then when a list gets passed in, it gets deduplicated automatically. Then you can override `json()` and `dict()` to convert it back to a list on the way out.

**Comment by uriyyo** [comment]
@mdavis-xyz The idea of this feature is to validate that sequences have only unique items.

Code below will show the idea:
```python
from typing import Set

from pydantic import BaseModel, Field


class FooForm(BaseModel):
    tags: Set[int]


foo = FooForm(tags=[1, 1, 2])  # tags attribute will be {1, 2}


class BarForm(BaseModel):
    tags: Set[int] = Field(unique=True)


bar = BarForm(tags=[1, 1, 2])  # will fail, because sequence has duplicate items
```

If you have any questions, feel free to ask.

**Comment by mpkocher** [comment]
This is an interesting issue that clarifies some of the friction points with validation and casting within Pydantic. 

As a workaround, it's possible to add a validator with 'pre=True' to get the desired behavior. 

```python
from typing import Set, Sequence
from pydantic import BaseModel, Field, validator

def custom_to_set(xs: Sequence[int]) -> Set[int]:
    items = set([])
    for item in xs:
        if item in items:
            raise ValueError(f"Duplicate item {item}")
        else:
            items.add(item)
    return items


class Record(BaseModel):
    tags: Set[int] = Field(...)

    @validator('tags', pre=True)
    def validate_unique_tags(cls, value):
        # raise ValueError or a set
        return custom_to_set(value)

def example() -> int:

    t1 = ['1', '2'] # Notice how liberal Pydantic is. This is not a validation error
    t2 = [1, 3, 1] # This now raises
    for t in [t1, t2]:
        print(Record(tags=t))
    return 0
```


Is this issue a bandaid for the fundamental loose casting/coercion model that Pydantic has adopted?

https://github.com/samuelcolvin/pydantic/issues/1098

Your proposal does seem reasonable and clarifies the default behavior  of `Set[T]` in a "strict" mode should raise a core validation error if the input sequence has any duplicates. However, this is not how Pydantic has approached validation.  Perhaps it would be better to wait until Pydantic has a coherent "strict" model?

With regards to the proposal, it's a bit counterintuitive to define a `Set[int]` and also set `Field(unique=True)` to get the validation to work "correctly".  Similarly, `List[T] = Field(unique=True)` is counterintuitive because it's not clear why a `Set[T]` isn't being used.  This case might be better solved by a custom validation/casting function with `pre=True`?

**Comment by uriyyo** [comment]
@mpkocher  This issue came up when I faced which unpredictable behavior I thought was a bug.

I had a model and one of the fields had a type `Set[str]`, the model looked like:
```python
class Form(BaseModel):
    tags: Set[int]
```
When one of the scripts tried to instantiate `Form` with duplicated items it simply cast sequence to a set, which was the wrong behavior for me script. After investigating I found that this is expected behavior, that why I create this PR)

I agree with you that `List[T] = Field(unique=True)` can be a little bit confusing.

Regarding #1098 it looks great and maybe it's better to return to this feature after `Strict Configuration` will be implemented.
